### PR TITLE
Implement mdns for `flutter run`

### LINKED ISSUE #40448: `flutter run` doesn't attach to observatory service on ios13
This is sort of related to [#37699](https://github.com/flutter/flutter/issues/37699). On an iPhone Xr device with ios 13, I am seeing very little device logs, and nothing from a flutter debug app. At first I thought this was a problem with `idevicesyslog`, but then I'm seeing the same thing from the macOS console app.

So I tried debugging flutter apps, running flutter apps, running non-flutter apps, changing system settings, and none of these produced any logs I could see in macOS's console app. The only logs I could get from the device where when I first attached it to my mac host: https://gist.github.com/christopherfujino/6e22a26d15c91fadd9feaf8fde97c5b4.

In an earlier version of ios, these syslogs are very busy.

My WIP PR to get `flutter run` to work using MDNS is [#40447](https://github.com/flutter/flutter/pull/40447).

## Logs

```
[✓] Flutter (Channel master, v1.10.2-pre.78, on Mac OS X 10.14.6 18G87, locale en-US)
    • Flutter version 1.10.2-pre.78 at /Users/fujino/git/flutter
    • Framework revision 066cad21a6 (28 minutes ago), 2019-09-13 11:05:00 -0700
    • Engine revision 7ea9884ab0
    • Dart version 2.5.0 (build 2.5.0-dev.4.0 be66176534)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/fujino/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
    • Xcode at /Users/fujino/Downloads/./Xcode.app/Contents/Developer
    • Xcode 11.0, Build version 11A419c
    • CocoaPods version 1.7.5

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 37.0.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] Connected device (1 available)
    • Christopher’s iPhone • 00008020-00017DA80CC1002E • ios • iOS 13.1

• No issues found!
```

### PR DESCRIPTION
**Pull request by christopherfujino**
## Description

Fixes the flutter tool not attaching to app observatory services on ios13 by implementing mDNS for `flutter run` (previously the tool would discover the observatory URI via reading the logs). If mDNS fails, the tool will fallback to the old method.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/40448

## Tests

I added the following tests:

Testing both mDNS discovery, and the fallback log scanning.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 1]
```diff
@@ -382,16 +382,36 @@ class IOSDevice extends Device {
...
+      Uri localUri;
       try {
         printTrace('Application launched on the device. Waiting for observatory port.');
-        final Uri localUri = await observatoryDiscovery.uri;
-        return LaunchResult.succeeded(observatoryUri: localUri);
+        localUri = await MDnsObservatoryDiscovery.instance.getObservatoryUri(
+          package.id,
+          this,
+          ipv6,
+          debuggingOptions.observatoryPort
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/devices.dart]: nit: missing comma

File: packages/flutter_tools/lib/src/mdns_discovery.dart [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  /// The [applicationId] parameter may be null, and can be used to
+  /// automatically select which application to use if multiple are advertising
+  /// Dart observatory ports.
+  MDnsObservatoryDiscovery({MDnsClient mdnsClient})
+    : client = mdnsClient ?? MDnsClient();
+
+  /// The [MDnsClient] used to do a lookup.
+  final MDnsClient client;
+
+  static const String dartObservatoryName = '_dartobservatory._tcp.local';
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/mdns_discovery.dart]: Does this need to be public? Could it be `@visibleForTesting`?

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 3]
```diff
@@ -382,16 +382,36 @@ class IOSDevice extends Device {
...
+        );
+        if (localUri != null) {
+          return LaunchResult.succeeded(observatoryUri: localUri);
+        }
+      } catch (error) {
+        printError('Failed to establish a debug connection with $id: $error');
+      }
+
+      // Fallback to manual protocol discovery
+      printTrace('mDNS lookup failed, attempting fallback protocol discovery.');
```
Comment (dnfield) [inline packages/flutter_tools/lib/src/ios/devices.dart]: nit: Maybe just clarify that this is going to try to scan logs in the message.

**Review by zanderso — APPROVED** [review]
lgtm w/ nits, but also wait for dnfield's review.

**Review by dnfield — APPROVED** [review]
This LGTM, would defer to @zanderso for final approval - in particular, do we want any analytics going on in here?  How often mDNS fails, how long it takes?

Also, can you update the pull request message template? :)

File: packages/flutter_tools/lib/src/mdns_discovery.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/mdns_discovery.dart]: added `@visibleForTesting`

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: done

File: packages/flutter_tools/lib/src/mdns_discovery.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/mdns_discovery.dart]: added `@visibleForTesting`

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: added comma

**Review by zanderso — APPROVED** [review]
still lgtm

**Pull request merged**